compiler/ballerina-lang/src/main/java/io/ballerina/projects/UsedBIRNodeAnalyzer.java

```diff
@@ -0,0 +1,685 @@
+    }
+
+    private void analyzeTestablePkg(BPackageSymbol testableSymbol) {
+        isTestablePkgAnalysis = true;
+        // since the testablePkg can access the nodes of parent package, we can merge the invocationData of both
+        // testable and parent packages
+        currentInvocationData.testablePkgInvocationData = testableSymbol.invocationData;
+
+        testableSymbol.invocationData.registerNodes(usedTypeDefAnalyzer, testableSymbol.bir);
+        // Analyzing testablePkg should be done first because it is the root of the dependency graph
+        for (BIRNode.BIRDocumentableNode node : testableSymbol.invocationData.startPointNodes) {
+            visitNode(node);
+        }
+        isTestablePkgAnalysis = false;
+    }
+
+    private void visitNode(BIRNode nodeToVisit) {
+        BIRNode.BIRFunction prevParentFunction = currentParentFunction;
+        List<BIRNonTerminator> prevInstructionArr = currentInstructionArr;
+
+        nodeToVisit.accept(this);
+
+        currentParentFunction = prevParentFunction;
+        currentInstructionArr = prevInstructionArr;
+    }
+
+    @Override
+    public void visit(BIRNode.BIRFunction birFunction) {
+        if (birFunction.getUsedState() == UsedState.USED) {
+            return;
+        }
+
+        birFunction.markAsUsed();
+        currentInvocationData.addToUsedPool(birFunction);
+        currentParentFunction = birFunction;
+
+        if ((birFunction.flags & Flags.NATIVE) == Flags.NATIVE) {
+            registerUsedNativeClassPaths(birFunction);
+        }
+
+        birFunction.basicBlocks.forEach(this::visitNode);
+        birFunction.parameters.forEach(param -> usedTypeDefAnalyzer.analyzeTypeDefWithinScope(param.type, birFunction));
+        usedTypeDefAnalyzer.analyzeTypeDefWithinScope(birFunction.returnVariable.type, birFunction);
+    }
+
+    @Override
+    public void visit(BIRNode.BIRServiceDeclaration serviceDeclaration) {
+        serviceDeclaration.listenerTypes.forEach(
+                listenerType -> usedTypeDefAnalyzer.analyzeTypeDefWithinScope(listenerType, serviceDeclaration));
+    }
+
+    @Override
+    public void visit(BIRNode.BIRTypeDefinition typeDef) {
+        if (typeDef.getUsedState() == UsedState.USED) {
+            return;
+        }
+
+        typeDef.markAsUsed();
+        currentInvocationData.addToUsedPool(typeDef);
+        usedTypeDefAnalyzer.analyzeTypeDef(typeDef);
+        typeDef.attachedFuncs.forEach(
+                attachedFunc -> addDependentFunctionAndVisit(typeDef, attachedFunc, typeDef.type.tsymbol.pkgID));
+    }
+
+    @Override
+    public void visit(BIRNode.BIRBasicBlock birBasicBlock) {
+        Map<BIRNode.BIRVariableDcl, FunctionPointerData> previousLocalFpHolders = localFpHolders;
+        currentInstructionArr = birBasicBlock.instructions;
+        localFpHolders = new HashMap<>();
+        localFpHolders.putAll(currentInvocationData.globalVarFPDataPool);
```

Comment:
Since the `localFPHolders` is being modified after this, initialized it with the size of the `currentInvocationData.globalVarFPDataPool`.  Fixed in f1840d75630f7f91c8848e721529c98216a65327.